### ISSUE TITLE
UoW Changesets from FlushEventArgs contain incorrect changes

### ISSUE THREAD

**Issue by followben** [mikro-orm/mikro-orm#864]
**Describe the bug**
On receiving an` afterFlush()` callback to our EntitySubscriber and accessing `args.uow.getChangeSets()`, we see UPDATE changes for entities we've populated during a `find()` but never mutated.

This occurs because `registerManaged()` in the UoW sometimes incorrectly receives WrappedEntity instances, so updates `__originalEntityData` with a reference, e.g. `{id: '12'}`. The changeset computer later diffs that with the full entity representation, sees it as a change and adds an UPDATE changeset to the persistStack.

On Martin's suggestion, I added `console.log(entity, new Error().stack)` to `registerManaged()` at L72, just after [it assigns the original entity data](https://github.com/mikro-orm/mikro-orm/blob/2b23912878ca8805312a28b38461acfe668d61c9/packages/core/src/unit-of-work/UnitOfWork.ts#L71). Our callsite is:

```typescript
await orm.em.getRepository(Credential).findOne(
	{
		// ... some where clause
	},
	['user.organisation']
);
```

Here's the output:

**Stack trace**
```
User {id: '1', __helper: WrappedEntity} Error: 
    at UnitOfWork.registerManaged (/Users/bstovold/Documents/backend/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js:60:33)
    at EntityFactory.create (/Users/bstovold/Documents//backend/node_modules/@mikro-orm/core/entity/EntityFactory.js:35:29)
    at EntityFactory.createReference (/Users/bstovold/Documents/node_modules/@mikro-orm/core/entity/EntityFactory.js:57:21)
    at ObjectHydrator.hydrateToOne (/Users/bstovold/Documents/backend/node_modules/@mikro-orm/core/hydration/ObjectHydrator.js:64:82)
    at ObjectHydrator.hydrateProperty (/Users/bstovold/Documents/backend/node_modules/@mikro-orm/core/hydration/ObjectHydrator.js:13:18)
    at ObjectHydrator.hydrate (/Users/bstovold/Documents/backend/node_modules/@mikro-orm/core/hydration/Hydrator.js:16:18)
    at EntityFactory.hydrate (/Users/bstovold/Documents/backend/node_modules/@mikro-orm/core/entity/EntityFactory.js:85:27)
    at EntityFactory.create (/Users/bstovold/Documents/backend/node_modules/@mikro-orm/core/entity/EntityFactory.js:33:14)
    at SqlEntityManager.findOne (/Users/bstovold/Documents/backend/node_modules/@mikro-orm/core/EntityManager.js:166:42)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at Function.Credential.fromApiKey (/Users/bstovold/Documents/backend/packages/database-entities/src/entities/credential.ts:116:22)
```

**To Reproduce**
I have not been able to reproduce this in a standalone test harness.

**Expected behavior**
Relations populated during `find()` but not altered should not end up in changesets.

**Additional context**
I'm upgrading from 4.0.0-rc2 to 4.6.0. This only occurs on 4.0.3 and above. On 4.0.0-rc3 to 4.0.2 inclusive I see a different problem with metadata references being undefined, which may or may still be there once this issue is resolved. The last functional release for us appears to be 4.0.0-rc2.

**Versions**

| Dependency | Version |
| - | - |
| node | 12.x |
| typescript | 3.9.3 |
| mikro-orm | 4.0.6 |
| pg | 3.8.3 |

**Comment by B4nan** [comment]
> This occurs because registerManaged() in the UoW sometimes incorrectly receives WrappedEntity instances

What you can see in the console log is not an instance of `WrappedEntity`, but a normal unloaded entity (= reference). Each entity have this `__helper: WrappedEntity`, but it's defined as non-enumerable, so it is quite weird that you can see it - my bet is that will be the main issue here. It looks like that entity was created from bad class - not discovered one. 

From the stack trace it also looks like the call begins in some entity, so how does that look like? Where do you get the EM from? Maybe you are using the wrong one, keep in mind that using the internal EM instance from entities is in general wrong, e.g. if you use explicit transaction demarcation, it will contain wrong EM.

Another observation - from the stack trace we can see that the path was "load entity, populate one of its to-one relations", but you end up with not populated reference, with just PK - so it was not found in the identity map. That again points to you using wrong EM instance, together with the above (call stack starts in entity), I am quite sure that is your problem.

**Comment by followben** [comment]
Yeah, `Credential.fromApiKey()` is just a static method which in turn calls findOne exactly as specified above. At this point we're not running in the context of an explicit transaction; the em is the one from our only orm instance.

> It looks like that entity was created from bad class - not discovered one

What do you mean by this?

**Comment by B4nan** [comment]
> Yeah, Credential.fromApiKey() is just a static method which in turn calls findOne exactly as specified above.

And where do you get the EM instance from that method? Are you sure it is called inside request context, not on the global EM? Try logging the `EntityManager.id` in that method and also in the `EntityFactory` where the `registerManaged` method is called to verify. 

> What do you mean by this?

I mean you might have multiple classes for the same entity, and only one is discovered. Because if the one you are using would be discovered, the `__helper` property would not be visible when `console.log`ing.

In the OP you said that the state is overridden, that also sounds wrong, at least from the stack trace you provided, I am quite sure there is no original entity data yet, as you can see, there is a `EntityFactory.createReference()` followed by `EntityFactory.create()` - that will never happen if that entity is loaded in the identity map. 

But again, if you are using wrong class reference, or wrong EM instance in general, this is exactly what would happen.

Without a reproduction this is not actionable, but from what you gave me, it points to a bug rather in your code than in the ORM... The fact that it was working before does not mean it was supposed to be working before :]

**Comment by B4nan** [comment]
Wait a minute, you said you cannot replicate it so I though it will be some weird edge case. But when I try to populate a relation, I can see pretty much the same as you mentioned, the original entity data are not set, only PK is there, which is definitely wrong. 

Really weird that things are working ok, and the only side effect of this is those update queries (but it kinda make sense in the end). The repro is as easy is this:

```ts
const god = new Author2('God', 'hello@heaven.god');
const bible = new Book2('Bible', god);
await orm.em.persistAndFlush(bible);
orm.em.clear();

const b = await orm.em.findOneOrFail(Book2, bible.uuid, ['author']);
expect(wrap(b.author, true).__originalEntityData).toMatchObject({ name: 'God', email: 'hello@heaven.god' });
```

**Comment by B4nan** [comment]
Ok, should be fixed now, will ship 4.0.7 right ahead. Let me know if it works fine now.

**Comment by followben** [comment]
Thanks @B4nan. Not sure what I missed in my original attempt at replicating standalone (it was very similar to yours), but was indeed the problem I was seeing... and 4.0.7 resolves it. 💥

**Comment by followben** [comment]
Actually... not so fast. Looks like diffing is still broken when custom types are involved. See my repro above @B4nan.

**Comment by B4nan** [comment]
Interesting, let me try again :]

**Comment by B4nan** [comment]
This looks more like an issue in your custom decimal type. When you look at the payload of the additional update changeset, you can see that it is trying to set the value to `'0'`, while it was loaded from db with `'0.0000'`. 

![image](https://user-images.githubusercontent.com/615580/94276314-5bbc0b80-ff48-11ea-880d-05e3796583af.png)

Looks like luxon will give you just `'0'`, but postgres will return it with all the defined decimals. You need to address that in the type definition, something like this works:

```ts
	convertToDatabaseValue(value: any, platform: Platform) {
    // eslint-disable-next-line eqeqeq
    if (value == null) { return value; }

		if (value instanceof Decimal) {
			if (value.isNaN()) { throw new Error('Decimal should not be NaN'); }
			if (!value.isFinite()) { throw new Error('Decimal should be finite'); }
      if (value.toString() === '0') { return '0.0000'; }

      return value.toString();
		}
		throw ValidationError.invalidType(DecimalType, value, 'JS');
	}
```

**Comment by followben** [comment]
The issue is not the custom type implementation but that the original entity data is populated with raw values from the DB, which is later compared with the converted to JS values and an update diff generated.

I included two examples in my repro: decimal.js for the n.000 and luxon for the datetime (notice in that payload the original aDateTime is incorrectly a string, not a luxon datetime object like it is in the entity data).

**Comment by B4nan** [comment]
> The issue is not the custom type implementation but that the original entity data is populated with raw values from the DB, which is later compared with the converted to JS values and an update diff generated.

Original entity data **should be** (and indeed are) populated with the raw values. And they are then compared with the output of the to-db method of the type.

**Comment by followben** [comment]
Then it's the latter that's not happening correctly. I've pushed a change to the repro at #871 that demonstrates this happens with Mikro's own custom types too (in this case [DateType](http://mikro-orm.io/docs/custom-types/#datetype)).

**Comment by B4nan** [comment]
That is a different story, postgres is apparently returning `Date` object for `date` columns automatically, which seems wrong/unexpected to me, so will force it to return strings there so this is consistent with mysql.

https://node-postgres.com/features/types#date--timestamp--timestamptz

**Comment by followben** [comment]
In that case then it looks like `BlobType` and `JsonType` are also broken @B4nan. See updated repro at #871.

**Comment by followben** [comment]
Also, if I need to patch in the number of decimal places in my custom type, how do I get hold of the column type defined alongside in the Property decorator? Our `DecimalType` provides a default on `getColumnType` of `numeric(12, 4)`, however I have properties that override this, e.g.

```
@Property({ type: DecimalType, columnType: 'numeric(8, 6)' })
latitude?: Decimal;
```

If the custom type needs to perform `toFixed()` during `convertToDatabaseValue()`, how can it get hold of the related entity metadata?

**Comment by B4nan** [comment]
Instead of overriding the `columnType`, you could pass the precision and scale to the type instance, e.g.

```ts
@Property({ type: new DecimalType(8, 6) })
```

Confirmed the provided json&blob types still have issues...

**Comment by followben** [comment]
Oh, I didn't realise you could pass instances to the Property type. I do think `compare` would be more flexible tho.

**Comment by B4nan** [comment]
Ok, now it should be fixed for good, without the need to handle the `0.0000` thing - as now when we hydrate a custom type, we also use the new value to update the entity data snapshot. 

Will be shipping 4.1 later this week, want to try few more performance related things before the release.
